Create a wrapper that checks for the presence of state modules for tests

Akm0d opened the pull request:
### What does this PR do?
Create a wrapper that checks for the presence of state modules for tests

s0undt3ch commented on tests/support/helpers.py:
```diff
@@ -1064,6 +1084,69 @@ def decorator(*args, **kwargs):
+    '''
+    Makes sure the passed salt state module is available. Skips the test if not
+
+    .. versionadded:: 0.5.2
```
Version added, 3000

Akm0d commented:
re-run full all

s0undt3ch commented on tests/integration/states/test_pkg.py:
```diff
@@ -28,12 +29,9 @@
-    pre_grains = LinuxDistribution()
-except ImportError:
-    pre_grains = None
+_grains = salt.loader.grains(build_minion_opts(minion_id='runtests-internal-sminion'))
```
Please, no dictionaries at the global scope.
If you need them to build the package list, maybe that should actually live in `setUpClass`?

s0undt3ch replied:
On top of being a dictionary, loading grains is not a blazing fast operation and will run just by importing the test module, this alone will slow down pytest test collecting.

Akm0d committed:
Added helpers for pkg tests

 - created requires_salt_state_module
 - corrected misuse of requires_salt_module to use the new function
 - eliminated pre_grains and loaded real grains before the test starts
 - Various fixes to get the pkg tests actually running as they were skipped before
 - Skipped tests when requisite packages weren't available

s0undt3ch commented on tests/integration/states/test_pkg.py:
```diff
@@ -29,73 +31,50 @@
 
 @destructiveTest
 class PkgTest(ModuleCase, SaltReturnAssertsMixin):
+    _PKG_EPOCH_TARGETS = []
```
There's really no reason why these class attribute aren't also set in `setUpClass` right?

Akm0d replied:
The `skipIf` decorator doesn't have access to `self`.  These need to be declared outside of `setUpClass` so that certain tests can be skipped if they are empty.

s0undt3ch replied:
Ok

s0undt3ch commented on tests/integration/states/test_pkg.py:
```diff
@@ -583,46 +562,52 @@ def test_pkg_015_installed_held(self):
         )
 
+        if versionlock_pkg and '-versionlock is not installed' in str(ret):
+            self.skipTest(str(ret) + '  `{}` is installed'.format(versionlock_pkg))
```
Single formatted string here too.

Akm0d committed:
Fix mac_brew_pkg returning dict instead of string

dwoz commented on tests/support/helpers.py:
```diff
@@ -1067,6 +1067,111 @@ def decorator(*args, **kwargs):
+    return decorator
+
+
+def requires_salt_state_modules(*names):
```
Can this be `requires_salt_states` instead?

dwoz commented on tests/support/helpers.py:
```diff
@@ -1067,6 +1067,111 @@ def decorator(*args, **kwargs):
+    .. versionadded:: 3000
+    '''
+
+    def _check_required_salt_state_modules(*required_salt_modules):
```
Does this have to be a nested helper function? I'd prefer something like this live at the top level of the module.

Akm0d replied:
Was just following the format of `requires_salt_modules`, they share quite a bit of code though and it certainly could sit at the top level

Akm0d committed:
Separated check for sminion attributes to top level

Akm0d committed:
Merge branch 'master' of https://github.com/saltstack/salt into requires_salt_state_modules

s0undt3ch commented on tests/support/helpers.py:
```diff
@@ -1067,50 +1067,135 @@ def decorator(*args, **kwargs):
+    return decorator
+
+
+def check_required_sminion_attributes(sminion_attr, *required_items):
```
I understand why Daniel wanted this as a top level function, specially now that its used on 2 different helpers, but let's prefix it with a `_`.
This way people aren't tempted to use it directly and will create the right `requires_salt_*` helper function.

Akm0d replied:
I noticed this function might be usable in conftest.py since it contains an identical code block there, but since helpers.py imports from unittest is that a good idea?  Maybe there is another place this can go so that it is usable by conftest.py

s0undt3ch replied:
For now, leave it here.
Once we go PyTest, we'll rearrange.

Akm0d committed:
Fixed undefined variable, hid internal helper function

Akm0d committed:
Merge branch 'master' of https://github.com/saltstack/salt into requires_salt_state_modules

Akm0d committed:
Merge branch 'master' of https://github.com/saltstack/salt into requires_salt_state_modules

Akm0d committed:
Merge branch 'master' into requires_salt_state_modules

Akm0d committed:
Merge branch 'master' into requires_salt_state_modules

Akm0d commented:
re-run full windows2019

Akm0d committed:
Merge branch 'master' into requires_salt_state_modules

terminalmage commented:
This PR caused several tests to always be skipped. I discovered this while writing another test earlier tonight. Details are in https://github.com/saltstack/salt/pull/56681, and I've fixed the problem there.

Just a friendly reminder that whenever changes are made to test skipping behavior, it's that much more important to closely check the results for the affected tests, on *all* platforms, and not to assume that everything is OK because tests are green. :smiley:
